pkg/storaged/resize.jsx

```diff
@@ -0,0 +1,365 @@
+    is_mounted_synch
+} from "./utils.js";
+import {
+    existing_passphrase_fields, init_existing_passphrase,
+    request_passphrase_on_error_handler
+} from "./crypto-keyslots.jsx";
+import {
+    dialog_open, SizeSlider, BlockingMessage, TeardownMessage,
+    init_active_usage_processes
+} from "./dialog.jsx";
+import { std_reply } from "./stratis-utils.js";
+
+const _ = cockpit.gettext;
+
+export function lvol_and_fsys_resize(client, lvol, size, offline, passphrase) {
+    let fsys;
+    let crypto_overhead;
+    let vdo;
+    let stratis_bdev;
+    const orig_size = lvol.Size;
+
+    const block = client.lvols_block[lvol.path];
+    if (!block)
+        return lvol.Resize(size, { });
+
+    const crypto = client.blocks_crypto[block.path];
+    if (crypto) {
+        const cleartext = client.blocks_cleartext[block.path];
+        if (!cleartext)
+            return;
+        fsys = client.blocks_fsys[cleartext.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(cleartext);
+        stratis_bdev = client.blocks_stratis_blockdev[cleartext.path];
+        crypto_overhead = crypto.MetadataSize;
+    } else {
+        fsys = client.blocks_fsys[block.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(block);
+        stratis_bdev = client.blocks_stratis_blockdev[block.path];
+        crypto_overhead = 0;
+    }
+
+    function fsys_resize() {
+        if (fsys) {
+            // HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1934567
+            //
+            // block_fsys.MountedAt might be out of synch with reality
+            // here if resizing the crypto container accidentally
+            // triggered an unmount.  Thus, we check synchronously
+            // whether or not we should be doing a offline resize or
+            // not.
+            //
+            // Another option for us would be to just mount the
+            // filesystem back if that's what we expect, to undo the
+            // bug mentioned above. But let's be a bit more passive
+            // here and hope the bug gets fixed eventually.
+            return (is_mounted_synch(client.blocks[fsys.path])
+                    .then(is_mounted => {
+                        // When doing an offline resize, we need to first repair the filesystem.
+                        if (!is_mounted) {
+                            return (fsys.Repair({ })
+                                    .then(function () { return fsys.Resize(size - crypto_overhead, { }) }));
+                        } else {
+                            return fsys.Resize(size - crypto_overhead, { });
+                        }
+                    }));
+        } else if (vdo) {
+            if (size - crypto_overhead > vdo.physical_size)
+                return vdo.grow_physical();
+            else if (size - crypto_overhead < vdo.physical_size)
+                return Promise.reject(_("VDO backing devices can not be made smaller"));
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19400-20230927-134545-de938f8a-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)